### PR TITLE
feat: transfer ownership of event to another user

### LINKED ISSUE #3012: Transfer the ownership of your events to another organizer.
We are implementing feature to provide ability to a user to delete his/her account. If the user is associated with any event, he/she must be given choice to transfer the events to other organizers associated with the event. (Just like in GitHub)

github

**Additional context**
Server issue: https://github.com/fossasia/open-event-server/issues/5890

### PR DESCRIPTION
**Pull request by shreyanshdwivedi**
Fixes #3012 

#### Short description of what this resolves:
We are implementing feature to provide ability to a user to delete his/her account. If the user is associated with any event, he/she must be given choice to transfer the events to other organizers associated with the event.

#### Changes proposed in this pull request:
 - Adds a settings option which will have option to transfer ownership of an event to other user
ezgif com-video-to-gif(2)

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The acceptance, integration, unit tests and linter pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by kushthedude** [comment]
@shreyanshdwivedi Just a suggestion , `Semantic-UI` provides various designing patterns for the UI header and UI Segments and more , See if you can use `Red Segments` for this section to make it more look-a-like to `Github Danger Zone`

**Comment by shreyanshdwivedi** [comment]
@uds5501 @kushthedude I might need some help here. Actually the `events-transfer-modal` is coming up as plain html rather than being a modal. Can you guys point me what am I missing here?
Screenshot_2019-05-27 Settings Apk19 Events Open Event

**Comment by kushthedude** [comment]
> @uds5501 @kushthedude I might need some help here. Actually the `events-transfer-modal` is coming up as plain html rather than being a modal. Can you guys point me what am I missing here?
> Screenshot_2019-05-27 Settings Apk19 Events Open Event

@shreyanshdwivedi Incorrect File Name is all the cause buddy 😆

**Comment by shreyanshdwivedi** [comment]
@kushthedude thanks for pointing out my silly mistake :rofl:

**Comment by shreyanshdwivedi** [comment]
@uds5501 the frontend PR is almost completed now. It only needs the endpoint now. I'm sending the event to be transferred and the email of the new organizer as payload. Please take a look and make the endpoint based on it. And please reference the server PR here

**Comment by uds5501** [comment]
@shreyanshdwivedi understood, will link the PR over here and carry on the server implementation accordingly

File: app/controllers/events/view/settings.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    },
+    transferEvent(event, data) {
+      this.set('isLoading', true);
+      let payload = {
+        'data': {
+          event,
+          'new-organizer': data.newOrganizerEmail
+        }
+      };
+      this.loader
```
Comment (uds5501) [inline app/controllers/events/view/settings.js]: `return this.loader`
Also, try and work out the agreed upon implementation!

Comment (shreyanshdwivedi) [reply app/controllers/events/view/settings.js]: @uds5501 did your backend code is ready? Do your backend code require `return this.loader` and is it working?

Comment (shreyanshdwivedi) [reply app/controllers/events/view/settings.js]: @uds5501 I tried this 
```js
return this.loader
        .patch('/transfer-ownership', JSON.stringify(payload))
```
but params in my network tab is empty. Can you please check?

Comment (uds5501) [reply app/controllers/events/view/settings.js]: @shreyanshdwivedi there must be an error. what does it say?

Comment (shreyanshdwivedi) [reply app/controllers/events/view/settings.js]: @uds5501 the frontend is now sending data to server :smile: 
```
data: {
    event : {…}
    new-organizer : s@gmail.com
}
```
The data contains event info and new organizer email.
Screenshot from 2019-06-04 18-03-41

Currenlty error being raised -
`The method is not allowed for the requested URL.`

Comment (uds5501) [reply app/controllers/events/view/settings.js]: @shreyanshdwivedi we need to send this payload to `events/<event_id>` Please see if it is working in that case

**Comment by uds5501** [comment]
@shreyanshdwivedi please change the payload to :

```js
        data: {
          attributes: {
            'new-organizer': data.newOrganizerEmail,
            'starts-at'    : event.startsAt,
            'ends-at'      : event.endsAt,
            'timezone'     : event.timezone
          },
          id: event.originalId, 
          type: 'event',
          identifier: event.id
        }

```
This passes the schema validation checks and I can work on the `before_patch` hook after this

File: app/controllers/events/view/settings.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    transferEvent(event, data) {
+      console.log(event);
+      this.set('isLoading', true);
+      let payload = {
+        data: {
+          attributes: {
+            'new-organizer': data.newOrganizerEmail
+          },
+          type: 'event',
+          identifier: event.id
```
Comment (uds5501) [inline app/controllers/events/view/settings.js]: while you are at updating the payload, fix spacing here for build to pass

**Comment by shreyanshdwivedi** [comment]
@uds5501 I've updated the PR. Also I've changed `new-organiser` to `new-owner`. Please update your server implementation accordingly

**Comment by mariobehling** [comment]
@shreyanshdwivedi Please also update the text about the account deletion: 

Your account currently cannot be deleted as active events and/or orders are associated with it. Before you can delete your account you you must transfer the ownership of your events to another organizer or cancel your events. If you have ticket orders stored in the system, please cancel your orders first too.

**Comment by shreyanshdwivedi** [comment]
@uds5501 @CosmicCoder96  This PR is completed now. I've implemented a role-invite for transferring event and opened a server PR for handling role invite for `owner` role - https://github.com/fossasia/open-event-server/pull/6048
These both PRs can be merged only after `owner` is added as a role on server via this PR - https://github.com/fossasia/open-event-server/pull/6042

File: app/controllers/events/view/settings.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        'isEventTransferModalOpen'        : false,
+        'isConfirmEventTransferModalOpen' : true,
+        'checked'                         : false,
+        currentInvite
+      });
+    },
+    transferEvent() {
+      this.set('isLoading', true);
+      this.currentInvite.set('roleName', 'owner');
+      this.currentInvite.save()
```
Comment (niranjan94) [inline app/controllers/events/view/settings.js]: `async-await` would make this less verbose

File: app/templates/components/modals/confirm-event-transfer-modal.hbs [hunk 4]
```diff
@@ -0,0 +1,27 @@
+<div class="header">
+  {{t 'You are transferring this event. This action cannot be undone.'}}
+  <div class="muted small text">
+    {{t 'All the event rights will be transferred to another user'}}
+  </div>
+</div>
+<div class="content">
+  <form class="ui {{if isLoading 'loading'}} form" id="event-transfer-form" autocomplete="off" {{action 'submit' on='submit' preventDefault=true}}>
```
Comment (CosmicCoder96) [inline app/templates/components/modals/confirm-event-transfer-modal.hbs]: use snake case for the id.

File: app/templates/components/modals/confirm-event-transfer-modal.hbs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    <div class="field">
+      {{ui-checkbox label=(t 'Please tick the box to agree and press TRANSFER') type="checkbox" checked=checked onChange=(action (mut checked))}}
+    </div>
+  </form>
+</div>
+<div class="actions">
+  <button type="button" class="ui black button" {{action 'close'}}>
+    {{t 'Cancel'}}
+  </button>
+  <button type="submit" form="event-transfer-form" class="ui red button {{if (not checked) 'disabled'}}">
```
Comment (CosmicCoder96) [inline app/templates/components/modals/confirm-event-transfer-modal.hbs]: Use snake case for id.

**Comment by shreyanshdwivedi** [comment]
@CosmicCoder96 I've made the required change. Please review

**Comment by shreyanshdwivedi** [comment]
@CosmicCoder96 I've made the required change. Please review

File: app/components/modals/confirm-event-transfer-modal.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+export default ModalBase.extend(FormMixin, {
+  actions: {
+    submit() {
+      this.onValid(() => {
+        this.sendAction('transferEvent');
+      });
+    },
+    close() {
+      if (!this.currentInvite.get('id')) {
```
Comment (CosmicCoder96) [inline app/components/modals/confirm-event-transfer-modal.js]: why the `.get `?

File: app/controllers/events/view/settings.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        'isConfirmEventTransferModalOpen' : true,
+        'checked'                         : false,
+        currentInvite
+      });
+    },
+    async transferEvent() {
+      this.set('isLoading', true);
+      this.currentInvite.set('roleName', 'owner');
+      await this.currentInvite.save()
+        .then(() => {
```
Comment (CosmicCoder96) [inline app/controllers/events/view/settings.js]: then with await?
Read docs of async, await.

**Comment by shreyanshdwivedi** [comment]
@CosmicCoder96 updated the async-await logic. Please review

File: app/controllers/events/view/settings.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        this.set('isLoading', true);
+        this.currentInvite.set('roleName', 'owner');
+        const currentInviteObj = await this.currentInvite.save();
+        if (isEqual(currentInviteObj, this.currentInvite)) {
+          this.setProperties({
+            'isConfirmEventTransferModalOpen' : false,
+            'checked'                         : false
+          });
+          this.notify.success(this.l10n.t('Owner Role Invite sent successfully.'));
+          this.set('isLoading', false);
```
Comment (CosmicCoder96) [inline app/controllers/events/view/settings.js]: this needs to be done only once after the catch block is over.

Comment (shreyanshdwivedi) [reply app/controllers/events/view/settings.js]: @CosmicCoder96 updated. Please review

**Comment by shreyanshdwivedi** [comment]
@CosmicCoder96 @niranjan94 please review

File: app/controllers/events/view/settings.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        'checked'                         : false,
+        currentInvite
+      });
+    },
+    async transferEvent() {
+      try {
+        this.set('isLoading', true);
+        this.currentInvite.set('roleName', 'owner');
+        const currentInviteObj = await this.currentInvite.save();
+        if (isEqual(currentInviteObj, this.currentInvite)) {
```
Comment (niranjan94) [inline app/controllers/events/view/settings.js]: Not required.

**Comment by shreyanshdwivedi** [comment]
@niranjan94 I've made the changes. Please review

File: app/components/modals/event-transfer-modal.js [hunk 10]
```diff
@@ -0,0 +1,10 @@
+import { computed } from '@ember/object';
+import ModalBase from 'open-event-frontend/components/modals/modal-base';
+
+export default ModalBase.extend({
+  isSmall          : true,
+  confirmEventName : '',
+  isNameDifferent  : computed('confirmEventName', function() {
```
Comment (niranjan94) [inline app/components/modals/event-transfer-modal.js]: You're using `eventName` too. So, you must ensure your computed property has that also as a dependant key.

**Comment by shreyanshdwivedi** [comment]
@niranjan94 added `eventName` to computed properties. Please check

**Comment by niranjan94** [comment]
@shreyanshdwivedi this PR contains `DO-NOT-MERGE` label. Is the label still valid ?

**Comment by shreyanshdwivedi** [comment]
Let's not merge it for now. It depends on PR https://github.com/fossasia/open-event-server/pull/6042
Please review it too @CosmicCoder96 @niranjan94
It is a big change so needs a thorough check

**Pull request merged**